provider/oracle/firewall.go

```diff
@@ -0,0 +1,659 @@
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	allIpLists := []ociResponse.SecIpList{}
+	for _, val := range secIpLists.Result {
+		allIpLists = append(allIpLists, val) //[val.Name] = val
+	}
+	for _, val := range defaultSecIpLists.Result {
+		allIpLists = append(allIpLists, val) //[val.Name] = val
+	}
+	return allIpLists, nil
+}
+
+func (f *Firewall) getAllIPListsAsMap() (map[string]ociResponse.SecIpList, error) {
+	allIps, err := f.getAllIPLists()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	allIpLists := map[string]ociResponse.SecIpList{}
+	for _, val := range allIps {
+		allIpLists[val.Name] = val
+	}
+	return allIpLists, nil
+}
+
+func (f *Firewall) isSecIpList(name string) bool {
+	if strings.HasPrefix(name, "seciplist:") {
+		return true
+	}
+	return false
+}
+
+func (f *Firewall) isSecList(name string) bool {
+	if strings.HasPrefix(name, "seclist:") {
+		return true
+	}
+	return false
+}
+
+func (f *Firewall) ensureApplication(portRange network.PortRange, cache *[]ociResponse.SecApplication) (string, error) {
+	for _, val := range *cache {
+		if val.PortProtocolPair() == portRange.String() {
+			return val.Name, nil
+		}
+	}
+	// We need to create a new application
+	// There is always the chance of a race condition
+	// when it comes to creating new resources.
+	// ie: someone may have already created a matching
+	// application between the time we fetched all of them
+	// and the moment we actually got to create one
+	// Worst thing that can happen is that we have a few duplicate
+	// rules, that we cleanup anyway when we destroy the environment
+	uuid, err := utils.NewUUID()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	secAppName := f.newResourceName(uuid.String())
+	var dport string
+	if portRange.FromPort == portRange.ToPort {
+		dport = strconv.Itoa(portRange.FromPort)
+	} else {
+		dport = fmt.Sprintf("%s-%s",
+			strconv.Itoa(portRange.FromPort), strconv.Itoa(portRange.ToPort))
+	}
+	name := f.client.ComposeName(secAppName)
+	secAppParams := oci.SecApplicationParams{
+		Description: "Juju created security application",
+		Dport:       dport,
+		Protocol:    ociCommon.Protocol(portRange.Protocol),
+		Name:        name,
+	}
+	application, err := f.client.CreateSecApplication(secAppParams)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	*cache = append(*cache, application)
+	return application.Name, nil
+}
+
+func (f *Firewall) ensureSecList(name string) (ociResponse.SecList, error) {
+	details, err := f.client.SecListDetails(name)
+	if err != nil {
+		if oci.IsNotFound(err) {
+			details, err := f.client.CreateSecList(
+				"Juju created security list",
+				name,
+				ociCommon.SecRulePermit,
+				ociCommon.SecRuleDeny)
+			if err != nil {
+				return ociResponse.SecList{}, nil
+			}
+			return details, nil
+		}
+	}
+	return details, nil
+}
+
+// TODO (gsamfira):finish this
+func (f *Firewall) ensureSecIpList(cidr []string, cache *[]ociResponse.SecIpList) (string, error) {
```

Comment:
I'm not so sure about this approach. How do you delete the these IP lists? I think it might be better to have a list per port-range, since each port range can have only one list of source CIDRs associated (it might change over time, though). Then if the source CIDRs change for a port-range, you can update the CIDRs in the sec IP list.